File: shipbob_to_gorgias/rust/src/gorgias.rs
//! Gorgias client, and the payloads both flows send it.
//!
//! The cron flow upserts customers, writing ShipBob order history to
//! `data.pandium.shipbob_orders`; the webhook flow creates tickets.
//!
//! Auth is OAuth2 via Pandium's `gorgias-oauth` connector. Pandium runs the
//! authorization flow when the tenant connects and handles refreshes, so this 
//! client never sees a client secret, never posts to a token endpoint, and 
//! holds no refresh logic 

use std::sync::LazyLock;
use std::time::Duration;

use anyhow::{Context, Result};
use regex::Regex;
use serde_json::{Value, json};

use crate::dates;
use crate::http;
use crate::pandium::Pandium;
use crate::shipbob::{Address, Recipient};

/// Gorgias validates the shape of an email before it will store it; mirror that
/// check so the two systems agree on which recipients get an email-keyed
/// customer and which fall back to a synthetic key.
static EMAIL: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(concat!(
        r"^([-!#-'*+/-9=?A-Z^-~]+(\.[-!#-'*+/-9=?A-Z^-~]+)*|",
        "\"([\\]!#-\\[^-~ \\t]|(\\\\[\\t -~]))+\")",
        r"@([-!#-'*+/-9=?A-Z^-~]+(\.[-!#-'*+/-9=?A-Z^-~]+)*|\[[\t -Z^-~]*])$",
    ))
    .expect("the email pattern is a literal")
});

/// Whether Gorgias would accept `email` as an email address.
pub fn valid_email(email: &str) -> bool {
    !email.contains(".@") && EMAIL.is_match(email)
}

/// How a Gorgias customer is identified.
///
/// A ShipBob recipient often has no usable email, so both flows fall back to a
/// synthetic key built from the recipient's name and address. Making that an
/// enum rather than a pair of optional arguments means there is always exactly
/// one key, and the lookup and the created record cannot disagree about it.
pub enum CustomerKey {
    Email(String),
    ExternalId(String),
}

impl CustomerKey {
    /// A valid recipient email when there is one, otherwise a synthetic
    /// `name address1 city country`. Both flows key on this, so a webhook ticket
    /// lands on the same record that carries the customer's order history.
    pub fn for_recipient(recipient: &Recipient) -> Self {
        if let Some(email) = recipient.email.as_deref().filter(|e| valid_email(e)) {
            return Self::Email(email.to_string());
        }
        let Address {
            address1,
            city,
            country,
        } = &recipient.address;
        Self::ExternalId(
            [
                recipient.name.as_deref(),
                address1.as_deref(),
                city.as_deref(),
                country.as_deref(),
            ]
            .map(Option::unwrap_or_default)
            .join(" "),
        )
    }

    /// The key as Gorgias stores it on the customer's `external_id`.
    pub fn as_str(&self) -> &str {
        match self {
            Self::Email(email) => email,
            Self::ExternalId(external_id) => external_id,
        }
    }

    /// The query parameter `GET /customers` looks the customer up by.
    fn query(&self) -> (&'static str, String) {
        match self {
            Self::Email(email) => ("email", email.to_lowercase()),
            Self::ExternalId(external_id) => ("external_id", external_id.clone()),
        }
    }
}

/// Body for `POST /customers` when the customer does not yet exist.
pub fn new_customer_payload(recipient: &Recipient, key: &CustomerKey) -> Value {
    let mut payload = json!({
        "name": recipient.name.as_deref().unwrap_or_default(),
        "external_id": key.as_str(),
        "data": {"pandium": {"shipbob_orders": []}},
    });
    if let CustomerKey::Email(email) = key {
        payload["email"] = json!(email);
    }
    payload
}

/// One shipment as the sidebar shows it: ShipBob's own fields, with the dates
/// made readable and a deep link back into ShipBob added.
fn sidebar_shipment(shipment: &Value) -> Value {
    let mut shipment = shipment.clone();
    if let Some(fields) = shipment.as_object_mut() {
        for field in ["estimated_fulfillment_date", "actual_fulfillment_date"] {
            if let Some(value) = fields.get(field).and_then(Value::as_str) {
                let display = dates::for_display(value);
                fields.insert(field.to_string(), display.into());
            }
        }
        let id = fields
            .get("id")
            .and_then(Value::as_i64)
            .map(|id| id.to_string())
            .unwrap_or_default();
        let url = format!("https://web.shipbob.com/App/Merchant/#/Orders/{id}/");
        fields.insert("url".to_string(), url.into());
    }
    shipment
}

/// The single order entry stored in `data.pandium.shipbob_orders`.
pub fn order_entry(order: &Value) -> Value {
    let shipments: Vec<Value> = order["shipments"]
        .as_array()
        .map(Vec::as_slice)
        .unwrap_or(&[])
        .iter()
        .map(sidebar_shipment)
        .collect();

    json!({
        "id": order["id"],
        "created_date": dates::for_display(order["created_date"].as_str().unwrap_or_default()),
        "purchase_date": dates::for_display(order["purchase_date"].as_str().unwrap_or_default()),
        "reference_id": order["reference_id"],
        "order_number": order["order_number"],
        "status": order["status"],
        "type": order["type"],
        "channel": order["channel"],
        "shipping_method": order["shipping_method"],
        "recipient": order["recipient"],
        "products": order["products"],
        "tags": order["tags"],
        "shipments": shipments,
    })
}

/// The slice of Gorgias the two flows depend on: find-or-create a customer,
/// write order history onto them, open a ticket.
pub trait Helpdesk {
    /// The customer's detail record — the list endpoint omits `data`, which is
    /// where the order history lives — or `None` if there is no such customer.
    fn find_customer(&self, key: &CustomerKey) -> Result<Option<Value>>;

    /// Create the customer and return their new id.
    fn create_customer(&self, payload: &Value) -> Result<i64>;

    fn update_customer(&self, id: i64, payload: &Value) -> Result<()>;

    fn create_ticket(&self, payload: &Value) -> Result<Value>;
}

pub struct Gorgias {
    http: http::Client,
}

impl Gorgias {
    pub fn new(pandium: &Pandium) -> Result<Self> {
        let token = pandium.require_secret("gorgias_oauth_access_token")?;
        let account = pandium.require_secret("gorgias_oauth_account")?;
        // The connector reports its own scheme; every current Gorgias token is a
        // bearer, but read it rather than assume it.
        let token_type = pandium
            .secret("gorgias_oauth_token_type")
            .unwrap_or("Bearer");

        let base_url = format!("https://{}.gorgias.com/api", account.to_lowercase());
        log::info!("Gorgias API base URL: {base_url}");

        Ok(Self {
            http: http::Client::new(
                base_url,
                format!("{token_type} {token}"),
                Duration::from_secs(2),
            ),
        })
    }
}

impl Helpdesk for Gorgias {
    fn find_customer(&self, key: &CustomerKey) -> Result<Option<Value>> {
        let (parameter, value) = key.query();
        log::info!("looking for gorgias customer by {parameter} {value}");

        // An email or external_id maps to at most one customer, so there is
        // nothing to paginate through.
        let found = self.http.get("/customers", &[(parameter, value)])?;
        let Some(row) = found["data"].as_array().and_then(|rows| rows.first()) else {
            log::info!("customer not found");
            return Ok(None);
        };

        let id = row["id"].as_i64().context("Gorgias customer has no id")?;
        log::info!("customer found: {id}");
        Ok(Some(self.http.get(&format!("/customers/{id}"), &[])?))
    }

    fn create_customer(&self, payload: &Value) -> Result<i64> {
        log::info!("creating new gorgias customer");
        let created = self.http.post("/customers", payload)?;
        created["id"]
            .as_i64()
            .context("Gorgias created a customer without an id")
    }

    fn update_customer(&self, id: i64, payload: &Value) -> Result<()> {
        log::info!("updating gorgias customer {id}");
        self.http.put(&format!("/customers/{id}"), payload)?;
        Ok(())
    }

    fn create_ticket(&self, payload: &Value) -> Result<Value> {
        log::info!("creating gorgias ticket");
        self.http.post("/tickets", payload)
    }
}

File: shipbob_to_gorgias/ruby/gorgias.rb
require 'json'
require 'uri'
require 'date'
require 'faraday'
require 'faraday/retry'

require_relative 'lib'

module Sb2Gorgias
  # Only a recipient email Gorgias would actually accept counts as valid.
  EMAIL_RE = %r{\A(?:([-!#-'*+/-9=?A-Z^-~]+(\.[-!#-'*+/-9=?A-Z^-~]+)*|"([\]!#-\[^-~ \t]|(\\[\t -~]))+")@([-!#-'*+/-9=?A-Z^-~]+(\.[-!#-'*+/-9=?A-Z^-~]+)*|\[[\t -Z^-~]*\]))\z}

  # ShipBob sends 7-digit fractional seconds; ISO 8601 parsing takes at most 6.
  ISO_LONG_FRACTION_RE = /(\.\d{6})\d+/
  OFFSET_SUFFIX_RE = /(Z|[+-]\d{2}:?\d{2})\z/

  # A human-readable timezone name (UTC, or UTC+HH:MM) derived from the ORIGINAL string's
  # trailing offset - not from a parsed DateTime, which always carries an offset (defaulting
  # to +00:00 when the source had none), so "did this value have an offset at all?" can only
  # be answered from the source text.
  def self.tzname_suffix(iso_str)
    m = OFFSET_SUFFIX_RE.match(iso_str)
    return '' unless m
    return 'UTC' if m[1] == 'Z'

    sign = m[1][0]
    digits = m[1][1..].delete(':')
    hh, mm = digits[0, 2].to_i, digits[2, 2].to_i
    return 'UTC' if hh.zero? && mm.zero?

    format('UTC%s%02d:%02d', sign, hh, mm)
  end

  # Formats a ShipBob timestamp for the customer sidebar. An unparseable value passes through
  # unchanged rather than becoming an error or an empty string.
  def self.format_date(value)
    return '' if value.nil? || value == ''

    trimmed = value.to_s.gsub(ISO_LONG_FRACTION_RE, '\1')
    begin
      parsed = DateTime.iso8601(trimmed)
    rescue ArgumentError, TypeError
      return value
    end

    tzname = tzname_suffix(trimmed)
    out = parsed.strftime('%d/%m/%Y %H:%M:%S')
    tzname.empty? ? out : "#{out} #{tzname}"
  end

  # Gorgias API client - customer sidebar sync and shipment ticket creation.
  #
  # Auth is OAuth2 bearer via gorgias_oauth_access_token/_account/_token_type (default Bearer).
  class GorgiasAPI
    LOGGER = Sb2Gorgias.new_logger('gorgias')

    def initialize(pandium)
      secrets = pandium.secrets
      token = secrets['gorgias_oauth_access_token']
      account = secrets['gorgias_oauth_account']
      if token.nil? || token.empty? || account.nil? || account.empty?
        raise 'PAN_SEC_GORGIAS_OAUTH_ACCESS_TOKEN and PAN_SEC_GORGIAS_OAUTH_ACCOUNT are required'
      end

      @api_url = "https://#{account.downcase}.gorgias.com/api"
      token_type = secrets['gorgias_oauth_token_type']
      token_type = 'Bearer' if token_type.nil? || token_type.empty?

      @conn = Faraday.new(url: @api_url) do |f|
        # POST isn't retried on a bare timeout - Gorgias may have already processed it, and
        # retrying could create a duplicate. It still retries on a retryable status code.
        f.request :retry, max: 6, interval: 2, backoff_factor: 2,
                           retry_statuses: [429, 502, 503, 504], methods: %i[get put],
                           retry_if: ->(_env, exception) { exception.is_a?(Faraday::RetriableResponse) }
        f.headers['Authorization'] = "#{token_type} #{token}"
        f.headers['Accept'] = 'application/json'
        f.headers['Content-Type'] = 'application/json'
        f.options.open_timeout = 10
        f.options.timeout = 30
        f.adapter Faraday.default_adapter
      end
    end

    # Prefers email over external_id when both are given. find_customer has NO rescue here -
    # raises raw, unlike create_customer/update_customer/create_ticket below, which all log
    # and reraise. This asymmetry is deliberate.
    def find_customer(email: nil, external_id: nil)
      return nil if (email.nil? || email.empty?) && (external_id.nil? || external_id.empty?)

      LOGGER.info("looking for Gorgias customer: #{email}, #{external_id}")
      # Only email is lowercased - external_id is an exact match, and lowercasing it here
      # wouldn't match the mixed case new_customer_payload actually stored.
      query = email && !email.empty? ? "email=#{URI.encode_www_form_component(email.downcase)}"
                                      : "external_id=#{URI.encode_www_form_component(external_id)}"
      res = @conn.get("customers?#{query}")
      raise "Gorgias customer lookup failed: #{res.status}" unless res.success?

      rows = JSON.parse(res.body)['data'] || []
      if rows.empty?
        LOGGER.info('customer not found')
        return nil
      end

      detail = @conn.get("customers/#{rows[0]['id']}")
      raise "Gorgias customer detail fetch failed: #{detail.status}" unless detail.success?

      LOGGER.info('customer found')
      JSON.parse(detail.body)
    end

    def create_customer(payload)
      LOGGER.info('creating new Gorgias customer')
      res = @conn.post('customers') { |r| r.body = JSON.generate(payload) }
      raise "Gorgias create customer failed: #{res.status} #{res.body}" unless res.success?

      LOGGER.info('customer created successfully')
      JSON.parse(res.body)['id']
    rescue Faraday::Error => e
      LOGGER.error("create customer failed: #{e}")
      raise
    end

    def update_customer(customer_id, payload)
      LOGGER.info("updating Gorgias customer #{customer_id}")
      res = @conn.put("customers/#{customer_id}") { |r| r.body = JSON.generate(payload) }
      raise "Gorgias update customer failed: #{res.status} #{res.body}" unless res.success?

      LOGGER.info('customer updated')
      nil
    rescue Faraday::Error => e
      LOGGER.error("update customer #{customer_id} failed: #{e}")
      raise
    end

    def create_ticket(payload)
      LOGGER.info('creating Gorgias ticket')
      res = @conn.post('tickets') { |r| r.body = JSON.generate(payload) }
      raise "Gorgias create ticket failed: #{res.status} #{res.body}" unless res.success?

      JSON.parse(res.body)
    rescue Faraday::Error => e
      LOGGER.error("create ticket failed: #{e}")
      raise
    end

    def valid_email(email)
      email && !email.empty? && !email.include?('.@') && EMAIL_RE.match?(email) ? email : ''
    end

    # Email when there is one, otherwise a synthetic "name address1 city country" key.
    def customer_key(sb_order)
      email = valid_email(Sb2Gorgias.deep_get(sb_order, 'recipient.email', ''))
      return email unless email.empty?

      address = Sb2Gorgias.deep_get(sb_order, 'recipient.address', {})
      [
        Sb2Gorgias.loose_or(Sb2Gorgias.deep_get(sb_order, 'recipient.name', ''), ''),
        Sb2Gorgias.loose_or(Sb2Gorgias.deep_get(address, 'address1', ''), ''),
        Sb2Gorgias.loose_or(Sb2Gorgias.deep_get(address, 'city', ''), ''),
        Sb2Gorgias.loose_or(Sb2Gorgias.deep_get(address, 'country', ''), ''),
      ].join(' ')
    end

    def new_customer_payload(sb_order, key)
      payload = {
        'name' => Sb2Gorgias.deep_get(sb_order, 'recipient.name', ''),
        'external_id' => key,
        'data' => { 'pandium' => { 'shipbob_orders' => [] } },
      }
      email = valid_email(Sb2Gorgias.deep_get(sb_order, 'recipient.email', ''))
      payload['email'] = email unless email.empty? # key entirely absent otherwise, not ''
      payload
    end

    def order_data_payload(sb_order)
      shipments = Sb2Gorgias.deep_get(sb_order, 'shipments', []) || []
      shipments.each do |shipment|
        %w[estimated_fulfillment_date actual_fulfillment_date].each do |field|
          shipment[field] = Sb2Gorgias.format_date(shipment[field]) if shipment[field]
        end
        shipment['url'] = "https://web.shipbob.com/App/Merchant/#/Orders/#{shipment['id']}/"
      end
      {
        'id' => Sb2Gorgias.deep_get(sb_order, 'id', ''),
        'created_date' => Sb2Gorgias.format_date(Sb2Gorgias.deep_get(sb_order, 'created_date', '')),
        'purchase_date' => Sb2Gorgias.format_date(Sb2Gorgias.deep_get(sb_order, 'purchase_date', '')),
        'reference_id' => Sb2Gorgias.deep_get(sb_order, 'reference_id', ''),
        'order_number' => Sb2Gorgias.deep_get(sb_order, 'order_number', ''),
        'status' => Sb2Gorgias.deep_get(sb_order, 'status', ''),
        'type' => Sb2Gorgias.deep_get(sb_order, 'type', ''),
        'channel' => Sb2Gorgias.deep_get(sb_order, 'channel', {}),
        'shipping_method' => Sb2Gorgias.deep_get(sb_order, 'shipping_method', ''),
        'recipient' => Sb2Gorgias.deep_get(sb_order, 'recipient', {}),
        'products' => Sb2Gorgias.deep_get(sb_order, 'products', []),
        'tags' => Sb2Gorgias.deep_get(sb_order, 'tags', []),
        'shipments' => shipments,
      }
    end
  end
end
